leader/leader.go

```diff
@@ -1,59 +1,47 @@
 package leader
 
 import (
-	"encoding/json"
-	"io/ioutil"
-	"net/http"
-	"os"
-)
-
-const defaultURL = "http://localhost:4040/"
+	"context"
+	"sync/atomic"
 
-type LeaderElector struct {
-	URL string
-}
+	"github.com/hashicorp/go-hclog"
+	operator_leader "github.com/operator-framework/operator-lib/leader"
+	"k8s.io/client-go/kubernetes"
+)
 
-type LeaderResponse struct {
-	Name string `json:"name"`
+type Elector interface {
+	// IsLeader returns whether this host is the leader
+	IsLeader() (bool, error)
 }
 
-// New returns a LeaderElector with the default service endpoint
-func New() *LeaderElector {
-	return &LeaderElector{
-		URL: defaultURL,
-	}
+type LeaderForLife struct {
+	isLeader atomic.Value
 }
 
-// NewWithURL returns a LeaderElector with a custom service endpoint URL
-func NewWithURL(URL string) *LeaderElector {
-	return &LeaderElector{
-		URL: URL,
-	}
+// New returns a Elector that uses the operator-sdk's leader for life elector
+func New(ctx context.Context, logger hclog.Logger, clientset kubernetes.Interface) *LeaderForLife {
+	le := &LeaderForLife{}
+	le.isLeader.Store(false)
+
+	go func() {
+		// This function blocks until this replica becomes the "leader", which
+		// means it creates a ConfigMap with an ownerref. Another replica can
+		// become the leader when the current leader replica stops running, and
+		// the Kubernetes garbage collector deletes the vault-k8s-leader
+		// ConfigMap.
+		err := operator_leader.Become(ctx, "vault-k8s-leader")
```

Comment:
If a node calls this with a leader already elected, does this return an error, nil, or blocks until context cancelation?

Reply:
It looks like it blocks forever until it can achieve leadership. The context passed in is used for k8s API calls, but I don't see it being checked for cancellation in Become(): https://github.com/operator-framework/operator-lib/blob/v0.4.1/leader/leader.go#L88